contracts/solidity/transaction/Transaction.sol

```diff
@@ -0,0 +1,54 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.20;
+
+import { Secondary } from "./Secondary.sol";
+
+contract Transaction {
+    string public message;
+    uint myInteger;
+    address secondaryAddr;
+    Secondary secondaryContract;
+    event MsgValue(uint256);
+
+    constructor(address secondaryAddress) {
+       secondaryAddr = secondaryAddress;
+       secondaryContract = Secondary(payable(secondaryAddress));
+    }
+
+    function checkGasleft() public view returns (uint256) {
+        return gasleft();
+    }
+
+    function getMessageData(uint integer, string memory inputMessage) public returns (bytes memory) {
+        message = inputMessage;
+        myInteger = integer;
+
+        return msg.data;
+    }
+
+    function getMessageSender() public view returns (address) {
+        return msg.sender;
+    }
+
+    function getMessageSignature() public pure returns (bytes4) {
+        return msg.sig;
+    }
+
+    function getMessageValue() public payable {
+        emit MsgValue(msg.value);
+    }
+
+    function getGasPrice() public view returns (uint256) {
+        return tx.gasprice;
+    }
+
+    function getTxOriginFromSecondary() public view returns (address) {
+        address secondaryOrigin = secondaryContract.getTxOrigin();
```

Comment:
The local `secondaryOrigin` variable is redundant. The method could be simplified to 
```solidity
function getTxOriginFromSecondary() public view returns (address) {
    return secondaryContract.getTxOrigin();
}
```